### ISSUE TITLE
Overlapping TypeVar Options X and Union[X, Y] Lead To Unexpected Type Errors When Constructing New Value

### ISSUE THREAD

**Issue by AaronJRubin** [microsoft/pyright#7065]
**Describe the bug**
When defining a decorator that takes a function whose return type is a TypeVar that includes overlapping options, Pyright complains if your decorator logic "copies" the value returned from the original function and that value could match multiple TypeVar options.

**Code or Screenshots**
Compare the below `decorate` decorator definition (which typechecks correctly)

```python
IntOrInts = TypeVar("IntOrInts", int, Iterable[int])

def decorate(func: Callable[..., IntOrInts]) -> Callable[..., IntOrInts]:
    @wraps(func)
    def wrapper(*args, **kwargs) -> IntOrInts:
        result = func(*args, **kwargs)
        if isinstance(result, Iterable):
            return tuple(result)
        else:
            return int(result)
    
    return wrapper
```

... with the following definitions that raise type errors:

Code:
```python
IntsOrUnionOfIntAndInts = TypeVar("IntsOrUnionOfIntAndInts", Iterable[int], Union[int, Iterable[int]])

def decorate_iterable_and_union(func: Callable[..., IntsOrUnionOfIntAndInts]) -> Callable[..., IntsOrUnionOfIntAndInts]:
    @wraps(func)
    def wrapper(*args, **kwargs) -> IntsOrUnionOfIntAndInts:
        result = func(*args, **kwargs)
        if isinstance(result, Iterable):
            # There's a type error here, but there shouldn't be
            return tuple(result)
        else:
            return int(result)
    
    return wrapper
```

Error:
```
Expression of type "tuple[int, ...]" cannot be assigned to return type "IntsOrUnionOfIntAndInts@decorate_iterable_and_union"
  Type "tuple[int, ...]" cannot be assigned to type "IntsOrUnionOfIntAndInts@decorate_iterable_and_union"
```

Code:
```python
IntOrUnionOfIntAndInts = TypeVar("IntOrUnionOfIntAndInts", int, Union[int, Iterable[int]])

def decorate_single_value_and_union(func: Callable[..., IntOrUnionOfIntAndInts]) -> Callable[..., IntOrUnionOfIntAndInts]:
    @wraps(func)
    def wrapper(*args, **kwargs) -> IntOrUnionOfIntAndInts:
        result = func(*args, **kwargs)
        if isinstance(result, Iterable):
            return tuple(result)
        else:
            # There's a type error here, but there shouldn't be
            return int(result)
    
    return wrapper
```

Error:
```
Expression of type "int" cannot be assigned to return type "IntOrUnionOfIntAndInts@decorate_single_value_and_union"
  Type "int" cannot be assigned to type "IntOrUnionOfIntAndInts@decorate_single_value_and_union"
```

Notably, if you just return the result as-is without "copying" it using the tuple or the int constructors, there is no type error here. I ran into this issue when trying to define a decorator using a TypeVar whose options were `int`, `Iterable[int]`, and `Union[int, Iterable[int]]`, but I was able to narrow things down a bit to the above minimal reproductions.

**VS Code extension or command-line**
Integrated into Pylance (v2023.12.1).

**Comment by erictraut** [comment]
I agree this is technically a bug, but it's an edge case that would be difficult to fix given the technique that pyright uses for tracking value-constrained TypeVars. See [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=value-constrained-type-variables) for details. I don't think it's worth trying to fix this edge case because your code is doing something that it really shouldn't in the first place.

Your definition of `IntOrUnionOfIntAndInts` includes two value constraints, but one of the constraint types (`Iterable[int]`) is a subtype of the other (`Iterable[int] | int`). That means the former type constraint is never applicable, and this TypeVar is effectively useless.

If your intent is to assure that the type is `int | Iterable[int]` or a subtype thereof, then you can simply use this union type rather than a TypeVar. If your use case is more complex than the sample would indicate and you really need a TypeVar, you can use a regular TypeVar with an upper bound of `Iterable[int] | int`.

If your intent is to assure that the type is constrained to either `int` or `Iterable[int]`, then you should use the `IntOrInts` TypeVar definition which uses non-overlapping types for its constraints.

**Comment by erictraut** [comment]
I presume one of the above suggestions is a reasonable workaround for you.

I'm going to close this as "won't fix".
